azerothcore/azerothcore-wotlk
# feat(InstanceScripts): new scriptable function DoAction()

This will work in similar way as UnitAI or GameObjectAI based DoAction function. This will allow to avoid SetData or ProcessEvent usage where basic input is needed.

This virtual fuction does not impacts current gameplay or similar, its used only for scriptable events where developer doesn't need to send any data, this works same as UnitAI::DoAction





## CHANGES PROPOSED:
-  Implements InstanceScript scriptable function "void DoAction(int32 actionId)"
-  


## ISSUES ADDRESSED:
- None



## TESTS PERFORMED:
- Build
- Tested ingame (AhnKahet rewrite on my fork)



## HOW TO TEST THE CHANGES:
- This virtual function doesn't impacts any script of currently scripted contents, this virtual function is invoked manually (same as UnitAI::DoAction), you must make calls inside of script or in other contents (gameobject/creature scripts) and add overriden DoAction function inside of specific instance script.

Call example (ACTION_TEST_ACTION needs to be defined for this test, it should be int32 (basic enum works too) type and any signed or unsigned value can be used):
```c
if (InstanceScript* instance = baseObject->GetInstanceScript())
    instance->DoAction(ACTION_TEST_ACTION);
```

Override example:
```c
void DoAction(int32 action) override
{
     if (action == ACTION_TEST_ACTION)
     {
         // You can make any particular actions here, this printf is just for example
         printf(">>>>>> ACTION_TEST_ACTION was invoked\n");
     }
}
```



- This can be tested only by developers, function usage is bit similar to InstanceScript::SetData calls, exception is that it does not require you to pass any data.

## KNOWN ISSUES AND TODO LIST:



## Target branch(es):








 
## How to test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here in the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR
